scripts/find-unused-fields.ts

```diff
@@ -0,0 +1,360 @@
+      rootNode.name.getStart()
+    )
+    if (!refs) {
+      throw new Error(
+        `Can't find references to fragment ${rootNode.name.getStart()}`
+      )
+    }
+
+    // filter out generated files
+    const fileNames = flatten(
+      refs.map(({ references }) => references.map(r => r.fileName))
+    ).filter(fn => !fn.endsWith(".graphql.ts"))
+
+    const searchString = `fragment ${fragmentName} on `
+    let loc: {
+      fileName: string
+      line: number
+      column: number
+    } = null
+    for (const fileName of fileNames) {
+      const lines = fs
+        .readFileSync(fileName)
+        .toString()
+        .split("\n")
+      const line = lines.findIndex(l => l.includes(searchString))
+      if (line < 0) {
+        continue
+      }
+      const column = lines[line].indexOf(searchString)
+      loc = {
+        fileName,
+        line: line + 1,
+        column: column + 1,
+      }
+      break
+    }
+
+    let path = relative(
+      process.cwd(),
+      (loc && loc.fileName) || sourceFile.fileName
+    )
+    if (loc) {
+      path += ":" + loc.line + ":" + loc.column
+    }
+
+    console.log("⚠️ ", chalk.bold(fragmentName), "→", chalk.cyan(path))
+    console.log(this.errors.map(line => "  " + line).join("\n"))
+    console.log()
+  }
+
+  // Recursively walk through the fragment's interface, checking each property's
+  // references as we go.
+  private checkTypeNode({
+    node,
+    path,
+  }: {
+    node: ts.TypeNode
+    path: ReadonlyArray<string>
+  }) {
+    if (ts.isParenthesizedTypeNode(node)) {
+      if (!node.type) {
+        console.error("bad node", node)
+        return
+      }
+      this.checkTypeNode({ node: node.type, path })
+      return
+    }
+    if (ts.isArrayTypeNode(node)) {
+      this.checkTypeNode({ node: node.elementType, path })
+      return
+    }
+    if (ts.isUnionTypeNode(node)) {
+      for (const child of node.types) {
+        this.checkTypeNode({ node: child, path })
+      }
+      return
+    }
+    if (ts.isTypeReferenceNode(node)) {
+      if (node.typeArguments) {
+        for (const child of node.typeArguments) {
+          this.checkTypeNode({ node: child, path })
+        }
+      }
+      return
+    }
+    if (!ts.isTypeLiteralNode(node)) {
+      switch (node.kind) {
+        case ts.SyntaxKind.AnyKeyword:
+        case ts.SyntaxKind.NumberKeyword:
+        case ts.SyntaxKind.BooleanKeyword:
+        case ts.SyntaxKind.StringKeyword:
+        case ts.SyntaxKind.NullKeyword:
+          break
+        default:
+          this.errors.push(
+            `Ignoring node of kind ${
+              ts.SyntaxKind[node.kind]
+            } at path ${path.join("/")}.`
+          )
+      }
+      return
+    }
+
+    for (const property of node.members) {
+      const propertyName = property.name.getText()
+      // ignore meta properties
+      if (
+        propertyName.startsWith('"') ||
+        propertyName === "__typename" ||
+        propertyName === "__id"
```

Comment:
I was slightly mistaken. The Global ID field will _always_ be selected, if available, the `__typename` field will only automatically be selected on union and interface types. If we can get away with only allowing the Global ID field, then that would probably be best.

Reply:
Isn't the global ID only on Node types while __typename is on everything?

Reply:
Not necessarily limited to types that explicitly implement the `Node` interface; but yes not all types might have a `__id` field, whereas `__typename` exists on all. I guess that does make it harder, we should just somehow tell people to prefer the ID field, when available. Perhaps that’s a GraphQL validation rule we can make (and also surface through vscode-apollo) 🤔